src/server/pps/server/api_server.go

```diff
@@ -3773,17 +3811,52 @@ func unknownError(ctx context.Context, msg string, err error) error {
 	return s.Err()
 }
 
+type cachedDefaultsGetter struct {
```

Comment:
why do we need this to be cached? it looks like it's only read once in the existing code path

Reply:
`SetClusterDefaults` ends up calling `apiServer.makeEffectiveSpec` once per pipeline; `SetProjectDefaults` will do the same for every pipeline in a project after CORE-1712.

Reply:
I’ve added some comments around defaults getting.